pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/BinaryOperatorTransformFunction.java

```diff
@@ -71,218 +113,331 @@ public TransformResultMetadata getResultMetadata() {
+      _results[i] = getIntResult((new ByteArray(leftBytesValues[i])).compareTo(new ByteArray(rightBytesValues[i])));
+    }
+  }
+
+  private void fillIntResultArray(ProjectionBlock projectionBlock, int[] leftIntValues, int length) {
+    int[] rightIntValues = _rightTransformFunction.transformToIntValuesSV(projectionBlock);
+    for (int i = 0; i < length; i++) {
+      _results[i] = getIntResult(Integer.compare(leftIntValues[i], rightIntValues[i]));
+    }
+  }
+
+  private void fillLongResultArray(ProjectionBlock projectionBlock, int[] leftValues, int length) {
+    long[] rightValues = _rightTransformFunction.transformToLongValuesSV(projectionBlock);
+    for (int i = 0; i < length; i++) {
+      _results[i] = getIntResult(Long.compare(leftValues[i], rightValues[i]));
+    }
+  }
+
+  private void fillFloatResultArray(ProjectionBlock projectionBlock, int[] leftValues, int length) {
+    float[] rightFloatValues = _rightTransformFunction.transformToFloatValuesSV(projectionBlock);
+    for (int i = 0; i < length; i++) {
+      _results[i] = getIntResult(Double.compare(leftValues[i], rightFloatValues[i]));
+    }
+  }
+
+  private void fillDoubleResultArray(ProjectionBlock projectionBlock, int[] leftValues, int length) {
+    double[] rightDoubleValues = _rightTransformFunction.transformToDoubleValuesSV(projectionBlock);
+    for (int i = 0; i < length; i++) {
+      _results[i] = getIntResult(Double.compare(leftValues[i], rightDoubleValues[i]));
+    }
+  }
+
+  private void fillStringResultArray(ProjectionBlock projectionBlock, int[] leftValues, int length) {
+    String[] rightStringValues = _rightTransformFunction.transformToStringValuesSV(projectionBlock);
+    for (int i = 0; i < length; i++) {
+      try {
+        _results[i] =
+            getIntResult(BigDecimal.valueOf(leftValues[i]).compareTo(new BigDecimal(rightStringValues[i])));
+      } catch (NumberFormatException e) {
+        _results[i] = 0;
+      }
+    }
+  }
+
+  private void fillIntResultArray(ProjectionBlock projectionBlock, long[] leftIntValues, int length) {
+    int[] rightIntValues = _rightTransformFunction.transformToIntValuesSV(projectionBlock);
+    for (int i = 0; i < length; i++) {
+      _results[i] = getIntResult(Long.compare(leftIntValues[i], rightIntValues[i]));
+    }
+  }
+
+  private void fillLongResultArray(ProjectionBlock projectionBlock, long[] leftValues, int length) {
+    long[] rightValues = _rightTransformFunction.transformToLongValuesSV(projectionBlock);
+    for (int i = 0; i < length; i++) {
+      _results[i] = getIntResult(Long.compare(leftValues[i], rightValues[i]));
+    }
+  }
+
+  private void fillFloatResultArray(ProjectionBlock projectionBlock, long[] leftValues, int length) {
+    float[] rightFloatValues = _rightTransformFunction.transformToFloatValuesSV(projectionBlock);
+    for (int i = 0; i < length; i++) {
+      _results[i] = getIntResult(Double.compare(leftValues[i], rightFloatValues[i]));
```

Comment:
We should use `BigDecimal` for this comparison

Reply:
I don't think this is necessary. `BigDecimal` is ridiculously expensive. Do you have a particular case in mind?

Reply:
I assume you are concerned about inexact translation to `double` - we now avoid converting to `BigDecimal` if the absolute value of the `long` is smaller than 2^53.